### PR TITLE
arch/sim: add windows host simulate support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arch/sim: add windows host simulate support
sim/host: move host implement to posix directory

This PR is depends on cmake support:
https://github.com/apache/incubator-nuttx/pull/6718

Preview version is on private project, will be contributed to the community asap:
```
git clone git@github.com:anchao/incubator-nuttx.git -b windows
git clone git@github.com:anchao/incubator-nuttx-apps.git -b windows
mv incubator-nuttx-apps apps
cd incubator-nuttx
```
Compile with cmake:
```
cmake -B vs2022 -DBOARD_CONFIG=sim/windows -G"Visual Studio 17 2022" -A Win32
cmake --build vs2022
```
If you want to use Visual Studio, you can import the solution directly from the IDE after CMake generates the solution:

6ba2d572-14e7-498f-96ea-4238c17d6803

After the compilation is completed, Debug\nuttx.exe will be generated:

c3e10a62-3e90-4dba-a6fd-64fb41f932ac

Run the simulator:

PowerShell:
(The default terminal operation has encoding issues, but it does not affect normal use, all commands can be executed normally)

b3c48de4-160c-4564-9aa3-c58902c3dbbd

Visual Studio

b83f4895-1ed9-43db-9fb3-0cf636ad2541

## Impact

N/A

## Testing

visual studio 2022 + cmake

### CODE REVIEW

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * Name: host_alloc_heap
+ *
+ * Description:
+ *   Allocate executable memory for heap.
+ *
+ ****************************************************************************/
+
+void *host_alloc_heap(size_t sz)
+{
+  return _aligned_malloc(sz, 8);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hostmemory.c:41]: HeapAlloc

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void *host_memalign(size_t alignment, size_t size)
+{
+  return _aligned_malloc(size, alignment);
+}
+
+void host_free(void *mem)
+{
+  _aligned_free(mem);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hostmemory.c:66]: HeapFree

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void *host_realloc(void *oldmem, size_t size)
+{
+  return _aligned_realloc(oldmem, size, 8);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hostmemory.c:71]: HeapReAlloc

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ ****************************************************************************/
+
+void *host_alloc_heap(size_t sz)
+{
+  return _aligned_malloc(sz, 8);
+}
+
+void *host_alloc_shmem(const char *name, size_t size, int master)
+{
+  return _aligned_malloc(size, 8);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hostmemory.c:46]: CreateFileMapping/MapViewOfFile
https://docs.microsoft.com/en-us/windows/win32/memory/creating-named-shared-memory

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void host_free_shmem(void *mem)
+{
+  _aligned_free(mem);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hostmemory.c:51]: UnmapViewOfFile

File: arch/sim/src/sim/win/up_hosttime.c [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: host_gettime
+ ****************************************************************************/
+
+uint64_t host_gettime(bool rtc)
+{
+  static LONGLONG start;
+  LARGE_INTEGER now;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:48]: FILETIME and remove the cast at line 50

File: arch/sim/src/sim/win/up_hosttime.c [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: host_gettime
+ ****************************************************************************/
+
+uint64_t host_gettime(bool rtc)
+{
+  static LONGLONG start;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:47]: uint64_t

File: arch/sim/src/sim/win/up_hosttime.c [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  LARGE_INTEGER due;
+  HANDLE timer;
+
+  /* Convert to 100 nanosecond interval,
+   * negative value indicates relative time
+   */
+
+  due.QuadPart = -(10 * ((nsec + 999) / 1000));
+
+  timer = CreateWaitableTimer(NULL, TRUE, NULL);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:83]: call Sleep instead

File: arch/sim/src/sim/win/up_hosttime.c [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+/****************************************************************************
+ * Name: host_gettime
+ ****************************************************************************/
+
+uint64_t host_gettime(bool rtc)
+{
+  static LONGLONG start;
+  LARGE_INTEGER now;
+
+  GetSystemTimeAsFileTime((LPFILETIME)&now);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:50]: let's use to void the time jump:
https://github.com/Alexpux/mingw-w64/blob/master/mingw-w64-libraries/winpthreads/src/clock.c

File: boards/sim/sim/sim/configs/windows/defconfig [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+# You can use "make menuconfig" to make any modifications to the installed .config file.
+# You can then do "make savedefconfig" to generate a new defconfig file that includes your
+# modifications.
+#
+CONFIG_ARCH="sim"
+CONFIG_ARCH_BOARD="sim"
+CONFIG_ARCH_BOARD_SIM=y
+CONFIG_ARCH_CHIP="sim"
+CONFIG_ARCH_FLOAT_H=y
+CONFIG_ARCH_SIM=y
```
[the remark this answers is not in the document]
Comment (anchao) [reply boards/sim/sim/sim/configs/windows/defconfig:13]: already define CONFIG_HOST_X86 in this config

File: arch/sim/src/Makefile [hunk 11]
```diff
@@ -24,6 +24,7 @@ ARCH_SRCDIR = $(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)$(DELIM)src
 
 INCLUDES += ${shell $(INCDIR) "$(CC)" $(ARCH_SRCDIR)}
 INCLUDES += ${shell $(INCDIR) "$(CC)" $(ARCH_SRCDIR)$(DELIM)chip}
+INCLUDES += ${shell $(INCDIR) "$(CC)" $(ARCH_SRCDIR)$(DELIM)chip$(DELIM)posix}
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/sim/src/Makefile]: Done

File: arch/sim/src/sim/win/up_hostmisc.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ ****************************************************************************/
+
+void host_abort(int status)
+{
+  PostQuitMessage(status);
+}
+
+int host_backtrace(void** array, int size)
+{
+  return 0;
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/sim/src/sim/win/up_hostmisc.c:48]: GOT!

File: arch/sim/src/sim/win/up_hostmemory.c [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+size_t host_malloc_size(void *mem)
+{
+  return _msize(mem);
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/sim/src/sim/win/up_hostmemory.c:56]: windows heap api does not provide the memalign version,  let us keep current implement

File: arch/sim/src/sim/win/up_hostmemory.c (hunk 5 above)
Comment (anchao) [reply arch/sim/src/sim/win/up_hostmemory.c:51]: Done

File: arch/sim/src/sim/win/up_hosttime.c (hunk 8 above)
Comment (anchao) [reply arch/sim/src/sim/win/up_hosttime.c:83]: Precision of Sleep() does not match with host api

File: arch/sim/src/Makefile [hunk 14]
```diff
@@ -75,7 +75,12 @@ CSRCS += up_vfork.c
 endif
 endif
 
-VPATH = sim
+VPATH = :sim
+ifeq ($(CONFIG_HOST_WINDOWS),y)
```
Comment (xiaoxiang781216) [inline arch/sim/src/Makefile:79]: CONFIG_WINDOWS_NATIVE

File: arch/sim/src/sim/win/up_hosttime.c [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define POW10_9 (1000000000ull)
+
+/* Number of 100ns-seconds between the beginning of the Windows epoch
+ * (Jan. 1, 1601) and the Unix epoch (Jan. 1, 1970)
+ */
+
+#define DELTA_EPOCH_IN_100NS  (0x19db1ded53e8000LL)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:39]: LL->ull

File: arch/sim/src/sim/win/up_hosttime.c [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: host_gettime
+ ****************************************************************************/
+
+uint64_t host_gettime(bool rtc)
+{
+  static uint64_t start;
+  LARGE_INTEGER freq, counter;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:52]: split to two line

File: arch/sim/src/sim/win/up_hosttime.c [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    {
+      QueryPerformanceFrequency(&freq);
+      QueryPerformanceCounter(&counter);
+
+      counter.QuadPart = counter.QuadPart * POW10_9 / freq.QuadPart;
+    }
+
+  if (rtc)
+    {
+      return counter.QuadPart;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:73]: move after line 61

File: arch/sim/src/sim/win/up_hosttime.c [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+  if (start == 0)
+    {
+      start = counter.QuadPart;
+    }
+
+  return counter.QuadPart - start;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:81]: move after line 68

File: tools/ci/testlist/sim-02.dat [hunk 19]
```diff
@@ -12,3 +12,6 @@
 
 # macOS doesn't have X11
 -Darwin,sim:touchscreen
+
+# Do not build Windows configs
+-Darwin,sim:windows
```
Comment (xiaoxiang781216) [inline tools/ci/testlist/sim-02.dat:17]: need disable linux too

File: arch/sim/src/sim/win/up_hosttime.c [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: host_gettime
+ ****************************************************************************/
+
+uint64_t host_gettime(bool rtc)
+{
+  static uint64_t start;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:51]: uint64_t to LARGE_INTEGER

File: arch/sim/src/sim/win/up_hosttime.c [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+/****************************************************************************
+ * Name: host_sleepuntil
+ ****************************************************************************/
+
+void host_sleepuntil(uint64_t nsec)
+{
+  uint64_t now;
+
+  now = host_gettime(false);
+  if (nsec > now + 1000)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:117]: if (nsec > now)

File: arch/sim/src/sim/win/up_hosttime.c [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+void host_sleep(uint64_t nsec)
+{
+  LARGE_INTEGER due;
+  HANDLE timer;
+
+  /* Convert to 100 nanosecond interval,
+   * negative value indicates relative time
+   */
+
+  due.QuadPart = -(10 * ((nsec + 999) / 1000));
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:97]:
```suggestion
  due.QuadPart = -((nsec + 99) / 100);
```

File: arch/sim/src/sim/win/up_hosttime.c [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+  LARGE_INTEGER freq, counter;
+  FILETIME ftime;
+
+  if (rtc)
+    {
+      GetSystemTimeAsFileTime(&ftime);
+
+      counter.QuadPart = (((uint64_t)ftime.dwHighDateTime << 32 |
+                          ftime.dwLowDateTime) -
+                          DELTA_EPOCH_IN_100NS) * 100;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_hosttime.c:61]: merge to previous linke

File: arch/sim/src/sim/win/up_hosttime.c (hunk 22 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/win/up_hosttime.c:97]: let's add macro NSTO100NS?

File: arch/sim/src/sim/win/up_simuart.c [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+/****************************************************************************
+ * Name: simuart_getc
+ ****************************************************************************/
+
+int simuart_getc(int fd)
+{
+  unsigned char ch;
+  DWORD nread;
+
+  if (ReadConsole(g_stdin_handle, &ch, 1, &nread, 0))
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_simuart.c:102]: check fd like simuart_checkc

File: arch/sim/src/sim/win/up_simuart.c [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+/****************************************************************************
+ * Name: simuart_putc
+ ****************************************************************************/
+
+int simuart_putc(int fd, int ch)
+{
+  DWORD nwritten;
+
+  if (WriteConsole(g_stdout_handle, &ch, 1, &nwritten, NULL))
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/win/up_simuart.c:85]: check fd like simuart_checkc

File: arch/sim/src/sim/up_internal.h [hunk 26]
```diff
@@ -32,7 +32,9 @@
 #ifndef __ASSEMBLY__
 #  include <sys/types.h>
 #  include <stdbool.h>
-#  include <netinet/in.h>
+#  if defined(CONFIG_SIM_NETDEV_TAP) || defined(CONFIG_SIM_NETDEV_VPNKIT)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_internal.h:35]: let's remove the empty vpnkit_ifup and vpnkit_ifdown, adn remove CONFIG_SIM_NETDEV_VPNKIT

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**